Fee Automation

xvi10 opened the pull request:
- Initial MultichainReader.sol, MultichainReaderUtils.sol and IOriginator.sol drafts
- Updates to Keys.sol and Errors.sol to support MultichainReader.sol
- EndpointV2Mock.sol for local hardhat testing (testing still needs to be configured)

SentinelDev1 committed:
Initial commit
- Initial MultichainReader.sol, MultichainReaderUtils.sol and IOriginator.sol drafts
- Updates to Keys.sol and Errors.sol to support MultichainReader.sol
- EndpointV2Mock.sol for local hardhat testing (testing still needs to be configured)

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -420,4 +420,16 @@ library Errors {
     error EmptyMarketPrice(address market);
+
+    // MultichainReader errors
+    error NotEnoughNative(uint256 msgValue);
```
rename to InsufficientMultichainNativeFee

SentinelDev1 replied:
Ok, no problem I can rename it as suggested. Just to note, most of these errors including this one were copied from the LayerZero code that was previously imported/inherited, but I don't see why we can't change the name on our end.

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -420,4 +420,16 @@ library Errors {
+
+    // MultichainReader errors
+    error NotEnoughNative(uint256 msgValue);
+    error LzTokenUnavailable();
```
where is this used?

SentinelDev1 replied:
This was used in the _payLzToken() function that was previously inherited from the OAppSender.sol contract. After refactoring to remove the various imports I removed because to minimize complexity, the MultichainReader contract is designed to pay using only the native token so it didn't seem necessary. If you agree using only native tokens to pay is sufficient, I will go ahead and remove this redundant error.

xvi10 replied:
i think using the native token to pay is fine

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -420,4 +420,16 @@ library Errors {
+    // MultichainReader errors
+    error NotEnoughNative(uint256 msgValue);
+    error LzTokenUnavailable();
+    error OnlyEndpoint(address addr);
```
should use Errors.Unauthorized instead

SentinelDev1 replied:
This is another error copied from the LayerZero contracts and thus using LayerZero naming, but I can update the naming as requested on our end.

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -420,4 +420,16 @@ library Errors {
+    error NotEnoughNative(uint256 msgValue);
+    error LzTokenUnavailable();
+    error OnlyEndpoint(address addr);
+    error ConfirmationsLengthMismatch(uint256 chainIdsLength, uint256 confirmationsLength);
```
where is this used?

SentinelDev1 replied:
This is an error I added earlier before using dataStore when number of confirmations per chain was being stored directly in the contract (the chain Ids and number of confirmations were array input params in the setter function - it was to make sure there wasn't a length mismatch between the two input arrays). Now it is being stored using the "MULTICHAIN_CONFIRMATIONS" key in dataStore thus is no longer used. I will remove it in the next commit.

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -420,4 +420,16 @@ library Errors {
+    error LzTokenUnavailable();
+    error OnlyEndpoint(address addr);
+    error ConfirmationsLengthMismatch(uint256 chainIdsLength, uint256 confirmationsLength);
+    error UnauthorizedOriginator(address originator);
```
should use Errors.Unauthorized instead

SentinelDev1 replied:
This is another error copied from the LayerZero contracts and thus using LayerZero naming, but I can update the naming as requested on our end.

xvi10 commented on contracts/external/MultichainReader.sol:
```diff
@@ -0,0 +1,230 @@
+        IOriginator(originator).processLzReceive(_guid, receivedData);
+        bytes memory transactionCallData = IOriginator(originator).transactionCallData(_guid);
+        if (transactionCallData.length != 0) {
+            (bool success, ) = originator.call(transactionCallData);
```
this doesn't seem safe, MultichainReader will have the CONTROLLER role, it should not call generic functions

`IOriginator(originator).processLzReceive(_guid, receivedData);` already seems sufficient, why is `originator.call(transactionCallData);` needed?

SentinelDev1 replied:
Are you suggesting the function in transactionCallData should be internal/private and invoked via processLzReceive in the originator contract? If so, internal functions can't be called with calldata the same way (via .call(calldata)), so the internal function would need to be explicitly defined in processLzReceive. This approach is fine but I think would require defining the function at the time of deploying the originator contract, reducing the flexibility of defining it dynamically during sendReadRequests(). Let me know if you agree with this alternate approach or have something else in mind.

xvi10 replied:
> Are you suggesting the function in transactionCallData should be internal/private and invoked via processLzReceive in the originator contract? 

yes

> If so, internal functions can't be called with calldata the same way (via .call(calldata)), so the internal function would need to be explicitly defined in processLzReceive

i think that is fine, i believe most originator contracts would just have one internal function to call on processLzReceive

if needed, the originator contract can have a mapping, e.g. mapping (bytes32 => ResponseHandler) guidHandlers

where ResponseHandler is an enum, and in processLzReceive the originator contract can determine the correct function to call based on the guid and the corresponding enum

> This approach is fine but I think would require defining the function at the time of deploying the originator contract, reducing the flexibility of defining it dynamically during sendReadRequests

i think we need to balance flexibility and security, even though there may not be an immediate risk to allowing the MultichainReader to call a generic function on the originator, and the originator contract should validate the caller for each of its functions, in case there is some missing access control, a combination of flexibility and development mistakes could lead to a compromise

minimizing access control over flexibility should be preferred where possible / practicle, especially where in this case the workaround of a mapping does not add too much overhead

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -420,4 +420,16 @@ library Errors {
+    error ConfirmationsLengthMismatch(uint256 chainIdsLength, uint256 confirmationsLength);
+    error UnauthorizedOriginator(address originator);
+    error OriginatorCallFailed(bytes transactionCallData);
+    error OnlyPeer(uint32 eid, bytes32 sender);
```
use Errors.Unauthorized instead

SentinelDev1 replied:
The first is our own error message, and given the concerns around originator.call(transactionCallData) it may no longer be necessary anyway. The second is another error using LayerZero naming but can also be changed on our end.

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -420,4 +420,16 @@ library Errors {
+    error UnauthorizedOriginator(address originator);
+    error OriginatorCallFailed(bytes transactionCallData);
+    error OnlyPeer(uint32 eid, bytes32 sender);
+    error NoPeer(uint32 eid);
```
rename to EmptyPeer

SentinelDev1 replied:
This is another error copied from the LayerZero contracts and thus using LayerZero naming, but I can update the naming as requested on our end.

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -420,4 +420,16 @@ library Errors {
+    error OriginatorCallFailed(bytes transactionCallData);
+    error OnlyPeer(uint32 eid, bytes32 sender);
+    error NoPeer(uint32 eid);
+    error InvalidEndpointCall();
```
where is this used?

SentinelDev1 replied:
It was in the IOAppCore interface which is no longer imported, I don't recall exactly where it was used but as it doesn't seem to be used I will remove it.

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -420,4 +420,16 @@ library Errors {
+    error OnlyPeer(uint32 eid, bytes32 sender);
+    error NoPeer(uint32 eid);
+    error InvalidEndpointCall();
+    error InvalidDelegate();
```
where is this used?

SentinelDev1 replied:
It was in the IOAppCore interface which is no longer imported, I don't recall exactly where it was used but as it doesn't seem to be used I will remove it.

xvi10 commented on contracts/external/MultichainReader.sol:
```diff
@@ -0,0 +1,230 @@
+    function sendReadRequests(
+        MultichainReaderUtils.ReadRequestInputs[] calldata readRequestInputs,
+        MultichainReaderUtils.ExtraOptionsInputs calldata extraOptionsInputs
+    ) external payable returns (MessagingReceipt memory, bytes32, MultichainReaderUtils.ReceivedData memory) {
```
might make sense to have an onlyController modifier here

SentinelDev1 replied:
Sorry, to clarify, is this referring to the setDelegate() function? 

I was thinking of removing it as previously discussed, but perhaps a safer option would be to use a highly restrictive modifier like onlyController. 

LayerZero responded to my questions around this on telegram saying that a delegate could be useful if faulty dvns need to be removed for example, even if other configuration changes aren't necessary.

xvi10 replied:
i'm referring to adding an onlyController modifier to sendReadRequests, because in lzReceive, the originator contract is called, and the originator contract would be validating that msg.sender is MultichainReader, so MultichainReader has some unique access to call potentially important contracts

so i think it would be safer to restrict access of who can call MultichainReader

SentinelDev1 committed:
Merge branch 'updates' into fee-automation

SentinelDev1 committed:
Updates related to PR comments and other minor updates

- Updated MultichainReader.sol per PR comments and other minor updates
- Updated Errors.sol per PR comments
- Removed unused function from IOriginator.sol
- Updated solidity version to 0.8.20 from 0.8.18 to support LayerZero imports

SentinelDev1 committed:
Added events and modifiers to MultichainReader.sol

SentinelDev1 committed:
Merge remote-tracking branch 'origin/updates' into fee-automation

SentinelDev1 committed:
Proposed implementation using lzReduce() to simplify the lzRead flow and logic

SentinelDev1 committed:
Simplified logic in _lzReceive()

SentinelDev1 committed:
Merge pull request #99 from gmx-io/fee-automation-lzreduce

Proposed implementation using lzReduce() to simplify the lzRead flow and logic

SentinelDev1 committed:
Removed readNumber from struct ReceivedData

SentinelDev1 committed:
Merge branch 'updates' into fee-automation

SentinelDev1 committed:
Initial updates to FeeDistributor

SentinelDev1 committed:
Further updates related to FeeDistributor

SentinelDev1 committed:
Various updates to FeeDistributor.sol and other related contracts

- Added multiple validations including checking if distribution already completed for the week, read response timestamp and to ensure FeeDistributor.distribute() is only run after FeeDistributor.initiateDistribute() is executed
- Added logic to dynamically decode the read responses depending on the number of chains
- Added initial logic to the FeeDistributor.distribute() function
- Related updates to Keys.sol and Errors.sol
- Other refactoring and general code improvements

SentinelDev1 committed:
Added initial logic for dynamic bridging calculations

SentinelDev1 committed:
Added initial bridging transaction logic and initial distribution calculation logic to the FeeDistributor.sol distribute() function

SentinelDev1 committed:
More updates to FeeDistributor.sol, Errors.sol and Keys.sol related to the fee calculations

xvi10 commented on contracts/data/Keys.sol:
```diff
@@ -467,6 +467,54 @@ library Keys {
+    // @dev key for FeeDistributor staked gmx
+    bytes32 public constant FEE_DISTRIBUTOR_STAKED_GMX = keccak256(abi.encode("FEE_DISTRIBUTOR_STAKED_GMX"));
+    // @dev key for FeeDistributor maximum bridging slippage
+    bytes32 public constant FEE_DISTRIBUTOR_MAX_SLIPPAGE = keccak256(abi.encode("FEE_DISTRIBUTOR_MAX_SLIPPAGE"));
```
could rename this to FEE_DISTRIBUTOR_MAX_BRIDGE_SLIPPAGE

SentinelDev1 replied:
No problem, this will be reflected in my next commit.

xvi10 commented on contracts/data/Keys.sol:
```diff
@@ -467,6 +467,54 @@ library Keys {
+    // @dev key for FeeDistributor number of chains with GMX deployments
+    bytes32 public constant FEE_DISTRIBUTOR_NUMBER_OF_CHAINS = keccak256(abi.encode("FEE_DISTRIBUTOR_NUMBER_OF_CHAINS"));
+    // @dev key for FeeDistributor chain ID
+    bytes32 public constant FEE_DISTRIBUTOR_CHAIN_ID = keccak256(abi.encode("FEE_DISTRIBUTOR_CHAIN_ID"));
```
i think we could use a chainId uint256 array in datastore instead of FEE_DISTRIBUTOR_NUMBER_OF_CHAINS, FEE_DISTRIBUTOR_CHAIN_ID

EDIT: i think you mentioned that you are already working on changing this

SentinelDev1 replied:
Yes, this was updated in my most recent commit.

xvi10 commented on contracts/data/Keys.sol:
```diff
@@ -467,6 +467,54 @@ library Keys {
+    // @dev key for FeeDistributor LayerZero version of chainId
+    bytes32 public constant FEE_DISTRIBUTOR_LAYERZERO_CHAIN_ID = keccak256(abi.encode("FEE_DISTRIBUTOR_LAYERZERO_CHAIN_ID"));
+    // @dev key for contract and keeper addresses used in FeeDistributor
+    bytes32 public constant FEE_DISTRIBUTOR_STORED_ADDRESSES = keccak256(abi.encode("FEE_DISTRIBUTOR_STORED_ADDRESSES"));
```
could rename this to FEE_DISTRIBUTOR_ADDRESS_INFO

SentinelDev1 replied:
No problem, this will be reflected in my next commit.

xvi10 commented on contracts/fee/FeeDistributor.sol:
```diff
@@ -4,199 +4,384 @@ pragma solidity ^0.8.0;
+            }
+
+            uint32 layerZeroChainId = uint32(dataStore.getUint(Keys.feeDistributorLayerZeroChainIdKey(chainId)));
+            uint256 readRequest = skippedCurrentChain ? (i - 1) * 3 : i * 3;
```
could rename this to readRequestIndex

SentinelDev1 replied:
No problem, this will be reflected in my next commit.

xvi10 commented on contracts/fee/FeeDistributor.sol:
```diff
@@ -4,199 +4,384 @@ pragma solidity ^0.8.0;
+                uint256 stakedGmx = IERC20(feeGmxTracker).totalSupply();
+                dataStore.setUint(Keys.feeDistributorFeeAmountKey(chainId), feeAmount);
+                dataStore.setUint(Keys.feeDistributorStakedGmxKey(chainId), stakedGmx);
+                skippedCurrentChain = true;
```
is skippedCurrentChain needed, or can the chainId array exclude the current chain?

SentinelDev1 replied:
I thought quite a bit about this as well, and we could take that approach, but decided against it for 2 reasons:

1. I felt that having the ChainId array consistent for the dataStore on each chain would be more manageable if and when updates are made, as the same data could be passed to each chain without having to take into account removing the current chain when populating or updating.
2. In order for the bridging logic on each chain to deterministically calculate the exact same amount of GMX bridged from one chain to another chain, I think it simplifies things to have the array the same on each chain.

SentinelDev1 committed:
Updated FeeDistributor.sol to add keeper costs calculation, correct array logic used and some other code cleanup

SentinelDev1 committed:
Updates to FeeDistributor.sol, Keys.sol, Errors.sol and added V1 interfaces

- Added the fee distribution logic to FeeDistributor.sol for sending out the calculated fees and costs
- Made changes related to comments from X to FeeDistributor.sol and Keys.sol
- Added buffer to the check of fee amounts on current chain to allow for the amount after bridging to be slightly less than the expected amount
- Various other updates to FeeDistributor
- The addition of IRewardDistributor.sol and IRewardTracker.sol interfaces

xvi10 commented on contracts/fee/FeeDistributor.sol:
```diff
@@ -4,199 +4,453 @@ pragma solidity ^0.8.0;
+                bridging[i] = new uint256[](chainCount);
+            }
+
+            uint256 deficit;
```
it is a bit difficult to understand the code here, i think that indexes should have the word index in the name to make it clearer, e.g. deficitIndex

also if there can be some comments that would help

i also think that the code could be re-written to be easier to understand, for example:
```
                while (deficit < chainCount && difference[deficit] >= 0) {
                    deficit++;
                }
                if (deficit == chainCount) break;
```
this seems to be moving the index forward until a chain with deficit is found, but the outer for loop is already looping through each chain, so it does not seem like moving the index forward within the loop is necessary

SentinelDev1 replied:
I reviewed the code again and added some comments yesterday to try to help make things a bit clearer and also added "Index" to the indexes as suggested.

SentinelDev1 committed:
Updates to FeeDistributor.sol and createdFeeDistributorVault.sol

xvi10 commented on contracts/error/Errors.sol:
```diff
@@ -420,4 +420,19 @@ library Errors {
+    error EmptyPeer(uint32 eid);
+
+    // FeeDistributor errors
+    error DistributionThisWeekAlreadyCompleted(uint256 lastDistributionTime, uint256 startOfCurrentWeek);
```
could rename this to "FeeDistributionAlreadyCompleted"

SentinelDev1 replied:
This has been updated.

xvi10 commented on contracts/fee/FeeDistributor.sol:
```diff
@@ -4,199 +4,467 @@ pragma solidity ^0.8.0;
+        );
+        feeDistributorVault.transferOut(gmxCurrentChain, extendedGmxTracker, feeAmountCurrentChain);
+
+        address gmxRewardDistributor = IRewardTracker(extendedGmxTracker).distributor();
```
there can be a function:
```
updateRewardDistributionRate(address tracker, amountForWeek) {
  address distributor = IRewardTracker(tracker).distributor();
  IRewardDistributor(distributor).updateLastDistributionTime()
  IRewardDistributor(distributor). setTokensPerInterval(amountForWeek / 1 weeks)
}
```

so that we do not need to remember to call updateLastDistributionTime for each call

SentinelDev1 replied:
I've updated this but also changed it a bit so that it takes address rewardToken as an input and instead of having:

`feeDistributorVault.transferOut(gmxCurrentChain, extendedGmxTracker, feeAmountCurrentChain);`

done separately before, it does this transfer as part of the internal function:

```
function updateRewardDistribution(address rewardToken, address tracker, uint256 rewardAmount) internal {
        feeDistributorVault.transferOut(rewardToken, tracker, rewardAmount);
        
        address distributor = IRewardTracker(tracker).distributor();
        IRewardDistributor(distributor).updateLastDistributionTime();
        IRewardDistributor(distributor). setTokensPerInterval(rewardAmount / 1 weeks);
    }
```

I've also updated the name to just updateRewardDistribution since it's also transferring the funds to the tracker.

SentinelDev1 replied:
Let me know if you are ok with these changes, prefer keeping the transfer separate, or have any other feedback.

SentinelDev1 committed:
Updates to FeeDistributor.sol, Keys.sol and Errors.sol

- Added functions for validating distribution states
- Added updateRewardDistribution function for repeated reward distribution code
- Reviewed bridging logic and added comments to make the logic more clear
- Removed FEE_DISTRIBUTOR_HAS_FEE_DEFICIT from Keys.sol as it is no longer used
- Updated naming of a few errors in Errors.sol

SentinelDev1 committed:
Updates to FeeDistributor.sol, Keys.sol and Errors.sol

- Many minor updates and optimizations in FeeDistributor.sol
- Chainlink fees are now calculated using a configurable factor in case the percentage that goes to chainlink were to change in the future
- Added a check to ensure that total bridged amounts from a chain with excess fees does not result in a deficit after bridging
- Added internal functions for dataStore.setUint, dataStore.getUint and dataStore.getAddress given that all are used frequently
- Removed check for if TotalStakedGmx = 0 given that it is not a realistic scenario in the context of this contract

xvi10 commented on contracts/fee/FeeDistributor.sol:
```diff
@@ -4,199 +4,522 @@ pragma solidity ^0.8.0;
+        eventData.boolItems.initItems(1);
+        eventData.boolItems.setItem(0, "distributeReferralRewards", false);
+
+        eventEmitter.emitEventLog("TriggerFeeDistributorKeeper", eventData);
```
i think instead of having the same event name TriggerFeeDistributorKeeper, and distributeReferralRewards be true / false

i think there should be different event names

e.g. for here the event name could be "FeeDistributionDataReceived"

in "distribute" the event name can be "FeeDistributionCompleted"

it would also be good for these events to provide more info on the fee distribution, e.g. how much GMX / WNT is distributed in total, how much WNT was sent to GLP, to keepers, treasury, Chainlink

SentinelDev1 replied:
I've just pushed an update with events updated as described above, though I named the first event "FeeDistributionReceivedData" instead of "FeeDistributionDataReceived" to better align the naming with other variable names used that reference receivedData

SentinelDev1 committed:
Updates to FeeDistributor.sol, Config.sol and Keys.sol

- Added FeeDistributor configuration keys to Config.sol (still need to add validations)

- Updated Keys.sol and FeeDistributor.sol per feedback received

SentinelDev1 committed:
Added referral rewards transfer to feeDistributionKeeper and updated variable names and events in FeeDistributor.sol

SentinelDev1 committed:
Updates to FeeDistributor.sol, Keys.sol and Errors.sol

- Added validation of esGMX for referral rewards and transfer of esGMX referral rewards to the FeeDistributionKeeper
- Cleanup of variable names, imports and addition of getAddress() overload helper function
- Cleanup of wnt referral rewards and treasury fee threshold errors and addition of esGMX referral rewards errors

SentinelDev1 committed:
Added GMX approval transaction necessary for bridging and refactored bridging transaction code in FeeDistributor.sol

SentinelDev1 committed:
Updates to FeeDistributor.sol, Errors.sol, Keys.sol and created IMintable.sol

- Refactored referral rewards logic checks and added two functions the keeper will use to send out referral rewards
- Added errors related to new referral rewards checks and functions to Errors.sol
- Added keys for storing the WNT price in USD and the amount of referral rewards sent out to Keys.sol
- Added IMintable.sol so the sendEsGmx() function can use the interface to mint more esGMX if necessary

SentinelDev1 committed:
Further updates related to FeeDistributor sendEsGmx() and sendWnt() functions

SentinelDev1 committed:
Updates to FeeDistributor.sol and EndpointV2Mock.sol

- various updates to fix stack too deep errors and lower contract size below the size limit
- added comments to various functions to improve code readability
- minor updates to EndpointV2Mock.sol to resolve compilation warnings

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Refactored FeeDistributor.sol to improve code readability and more efficiently resolve stack too deep errors

- Removed all but one struct and added various internal functions to improve code readability and resolve stack too deep errors without the need for many large structs

- Refactored GMX bridging calculation logic and added comments in an attempt to simplify the logic and improve readability

- Various other minor updates related to resolving stack too deep errors and improving code readability

SentinelDev1 committed:
Updated FeeDistributor.sol to update esGMX vester bonus rewards amounts when sending esGMX referral rewards

SentinelDev1 committed:
Various updates to FeeDistributor.sol, added required packages and reverted changes to Config.sol

- Updated FeeDistributor.sol to retrieve the chainlink GMX price and significant refactoring primarily to resolve stack and contract size issues
- Added needed packages to package.json
- Reverted inadvertant updates to Config.sol and Errors.sol that occured when changing the base branch

SentinelDev1 committed:
Further correction to Config.sol

SentinelDev1 committed:
Minor update to align hardhatconfig.ts with base branch

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Adding layerzero libraries back that were inadvertantly removed when merging updates from base branch

SentinelDev1 committed:
Added missing layerzero dependency

SentinelDev1 committed:
Updated Config.sol allowedBaseKeys for FeeDistributor.sol and MultichainReader.sol and added missing EXCLUDED_CONFIG_KEYS to config.ts

SentinelDev1 committed:
Updates to some comments in FeeDistributor.sol

SentinelDev1 committed:
Added supporting scripts for the soon to be added MultichainReader test script and updates to existing scripts

SentinelDev1 committed:
Contract size related optimizations in FeeDistributor.sol and moved MultichainReader, MultichainReaderUtils and IOriginator to contracts/multichain

SentinelDev1 committed:
Further optimizations to FeeDistributor.sol to decrease contract size and further updates to various scripts related to the MultichainReader tests

SentinelDev1 committed:
Minor correction to keys.ts

SentinelDev1 committed:
Initial MultichainReader.ts test script and updates to related contracts and scripts

SentinelDev1 committed:
Updates to the bridging flow in FeeDistributor.sol and further enhancements to MockEndpointV2.sol and the MultichainReader.ts test

SentinelDev1 committed:
Additional updates to the bridging logic to further simplify the calculations and optimize the code

SentinelDev1 committed:
Further optimizations and code improvements in FeeDistributor.sol

SentinelDev1 committed:
Updated FeeDistributor.sol to use V2 oracle WNT price and updated DataReceived event

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Corrected/added necessary excluded keys for testing

SentinelDev1 committed:
Further refinement of the updated FeeDistributor flow

SentinelDev1 committed:
Updated MultichainReader test script and added other scripts related to the feeDistributor test script

SentinelDev1 committed:
Updates to config.ts and keys.ts for the FeeDistributor.ts test script

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Added the initial draft of the FeeDistributor.ts test script and various other related updates

SentinelDev1 committed:
Updates to the FeeDistributor.ts test script and related files

SentinelDev1 committed:
Further updates to FeeDistributor test script and related files

SentinelDev1 committed:
Minor addition to keys.ts

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Attempt to correct unintended changes during base branch merge

SentinelDev1 committed:
Updates to FeeDistributor.ts test script and minor correction to FeeDistributor.sol

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Created scripts related to gelato keeper simulation plus FeeDistributor.ts and MultichainReader.ts test cleanup

SentinelDev1 committed:
Updates for the change from Synapse to Layerzero for cross-chain bridging of GMX

SentinelDev1 committed:
Updates to scripts related to the change in FeeDistributor.sol to use Layerzero bridging

SentinelDev1 committed:
Updates primarily related to the FeeDistribution keeper simulation scripts

SentinelDev1 committed:
Updates related to FeeDistributor.sol

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Minor corrections to files that shouldn't have changed.

SentinelDev1 committed:
More very minor corrections to Config.sol

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Minor correction to FeeDistributor.sol for recent merge changes

SentinelDev1 committed:
Minor update to scripts used for simulating the feeDistribution keeper

xdev10 committed:
update contributor handler

SentinelDev1 committed:
Initial updates related to the contributor payment automation

SentinelDev1 committed:
Updated bridging logic, added chainIds sorting in FeeDistributor, created FeeDistributorUtils library and updated FeeDistributor.ts test script

SentinelDev1 committed:
Minor corrections to FeeDistributor.ts test script

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Minor updates to FeeDistributor.ts

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Updates primarily related to renaming of FeeDistributor errors, keys and local variables

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Updated ContributorHandler.sol, created ContributorHandler deployment and test scripts and other minor test script updates

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Updates to FeeDistributor and related contracts and scripts for various updates related to fee calculations

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Further refinements to the FeeDistributor contract primarily related to excess V1 keeper costs

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Updates related to FeeDistributor GLP deficit logic and tests

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Update to simplify FeeDistributor logic and also avoid potential underflow when wntForReferralRewards is greater than maxKeeperAndReferralCostsGlp

SentinelDev1 committed:
Minor correction in FeeDistributor contract and added test to FeeDistributor.ts for WNT from Treasury

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Updates related to oracle price retrieval in the FeeDistributor setTokenPrices() function

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Created Keys2.sol for MultichainReader and FeeDistributor keys

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

xdev10 committed:
Merge branch 'v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Updates for FeeDistributor to use the claim flow and to alter the logic given V1 fees no longer sent to GLP

SentinelDev1 committed:
Additional updates related to using the claim flow and V1 changes

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Fix tests and minor correction to MockFeeDistributor.sol

SentinelDev1 committed:
Fix MultichainReader test

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Updates to FeeDistributor address retrieval functions and related changes

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Additional updates related to FeeDistributor address retrieval functions

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Updates related to recent oracleProviderForTokenKey changes

SentinelDev1 committed:
Now using MockGMX_Adapter contract for FeeDistributor.ts tests to validate compatibility with the new GMX bridging contracts

SentinelDev1 committed:
Updates to various scripts related to using MockGMX_Adapter contract for GMX bridging

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Minor correction to the feeDistribution keeper simulation scripts

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Updates to bring MockGMX_Adapter up to date with the contract post audit

SentinelDev1 committed:
Updates related to setting 2 ContributorHandler functions via TimelockConfig and renaming CONTRIBUTOR_TOKEN_VAULT and corresponding functions to CONTRIBUTOR_FUNDING_ACCOUNT

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

SentinelDev1 committed:
Updates remediating L-02 and I-02 fee automation audit findings

SentinelDev1 committed:
Additional updates to TimelockConfig related to finding L-03 in the fee-automation audit

xdev10 committed:
update script

SentinelDev1 committed:
Updates to the FeeDistributor calculateChainlinkAndTreasuryAmounts function chainlinkTreasuryWntAmount calculation and related changes

SentinelDev1 committed:
Updated FeeDistributor and MockFeeDistributor internal functions to use _ prefix in naming

xdev10 committed:
add custom contributor funding account

SentinelDev1 committed:
Updates primarily related to reducing contract size in TimelockConfig and some additional minor contract optimizations

SentinelDev1 committed:
Merge remote-tracking branch 'origin/v2.2-branch' into fee-automation

xdev10 committed:
update SendContributorPayment event

SentinelDev1 committed:
Add comments to FeeDistributor related to findings I-08 and I-09

SentinelDev1 committed:
Merge remote-tracking branch 'origin/updates' into fee-automation

xdev10 committed:
add note about FeeDistributor deployment

SentinelDev1 committed:
Merge remote-tracking branch 'origin/updates' into fee-automation

SentinelDev1 committed:
Merge remote-tracking branch 'origin/updates' into fee-automation

SentinelDev1 committed:
Updates to FeeDistributor.sol optimizing event emissions and ensuring efficient event filtering in the feeDistribution keeper

SentinelDev1 committed:
Additional minor updates to FeeDistributor related to the event emissions

SentinelDev1 committed:
Additional refinements of FeeDistributor bridging code related to LayerZero OFT contracts

SentinelDev1 committed:
Updated related to FeeDistributor.depositReferralRewards esGMX mint calculations

SentinelDev1 committed:
Added base sepolia to hardhat.config.ts and FeeDistributor deployment and config testnet scripts

SentinelDev1 committed:
Merge remote-tracking branch 'origin/updates' into fee-automation

SentinelDev1 committed:
Updated comment in FeeDistributor._bridgeGmx and minor updates to feeDistributorConfigTestnet.ts script

xdev10 committed:
add option to accept terms without a separate signing action

xdev10 committed:
fix tests

SentinelDev1 committed:
Removed unnecessary withOraclePrices modifier from _setTokenPrices() and removed _retrieveSetOraclePricesParams() function from FeeDistributor as well as related updates to various scripts
